Add documentation about apply napari style to separate windows (#4780)
Napari/napari: docs/plugins/best_practices.md

   from coverage with the comment `# pragma: no cover`
5. In the cookiecutter, coverage tests from github actions will be uploaded to
   codecov.io

## Set style for additional windows in your plugin

> In napari plugins we strongly advise to use widget docked in napari window,

Reviewer: ```suggestion
In napari plugins we strongly advise additional widgets be docked in the main napari viewer,
```

> but sometimes is required to use separate window. 

Reviewer: ```suggestion
but sometimes a separate window is required. 
```

The best practice is to use [`QDialog`](https://doc.qt.io/qt-5/qdialog.html)
based windows with parent set to widget

> already docked in napari window (sample code snip on the end).

Reviewer: ```suggestion
already docked in the viewer (sample code snippet at the end).
```

```python
from qtpy.QtWidgets import QDialog, QWidget, QSpinBox, QPushButton, QGridLayout, QLabel

class MyInputDialog(QDialog):
    def __init__(self, parent: QWidget):
        super().__init__(parent)
        self.setWindowTitle("My Input Dialog")
        self.number = QSpinBox()
        self.ok_btn = QPushButton("OK")
        self.cancel_btn = QPushButton("Cancel")
        
        layout = QGridLayout()
        layout.addWidget(QLabel("Number:"), 0, 0)
        layout.addWidget(self.number, 0, 1)
        layout.addWidget(self.ok_btn, 1, 0)
        layout.addWidget(self.cancel_btn, 1, 1)
        self.setLayout(layout)
        
        self.ok_btn.clicked.connect(self.accept)
        self.cancel_btn.clicked.connect(self.reject)
        
class MyWidget(QWidget):
  def __init__(self, viewer: "napari.Viewer"):
        super().__init__()
        self.viewer = viewer
        self.open_dialog = QPushButton("Open dialog")
        self.open_dialog.clicked.connect(self.open_dialog_clicked)
        
    def open_dialog(self):
        dialog = MyInputDialog(self)  

>         # this setting parent to self allow to inheritance of style from napari window

Reviewer: ```suggestion
        # setting parent to self allows the dialog to inherit its style from the viewer
```

I would probably move the comment above the line assigning dialog, so it's clear which line we're referring to (unless it's actually on the same line in which case ignore me sorry)

        dialog.exec_()
        if dialog.result() == QDialog.Accepted:
            print(dialog.number.value())
```

> When you want to use magicgui to create widget you could still wrap it in 

Reviewer: ```suggestion
When you want to use magicgui to create your widget you could still wrap it in 
```

`QDialog`:

```python
from typing import Callable

from magicgui import magicgui
from qtpy.QtWidgets import QDialog, QWidget, QVBoxLayout, QPushButton

def sample_add(a: int, b: int) -> int:
    return a + b

class MguiDialog(QDialog):
    def __init__(self, fun: Callable, parent=None):
        super().__init__(parent)
        self.mgui_widget = magicgui(fun)  # close of dialog will destroy widget
        layout = QVBoxLayout()
        layout.addWidget(self.mgui_widget.native)
        self.setLayout(layout)
        self.mgui_widget.called.connect(self.run)

    def run(self, value):
        print('run', value)
        self.close()

class MyWidget(QWidget):
    def __init__(self):
        super().__init__()
        self.btn = QPushButton('Click me')
        self.btn.clicked.connect(self.click)
        self.layout = QVBoxLayout()
        self.layout.addWidget(self.btn)
        self.setLayout(self.layout)

    def click(self):
        dialog = MguiDialog(sample_add, self)
        dialog.exec_()
```

> But if there are any reason that you need to use separate window that

Reviewer: ```suggestion
If there is a particular reason that you need to use a separate window that
```

> is `QWidget`, not `QDialog`, then you could use the `get_current_stylesheet` 

Reviewer: ```suggestion
inherits from `QWidget`, not `QDialog`, then you could use the `get_current_stylesheet` 
```

> and [`get_stylesheet`](/api/napari.qt.html#napari.qt.get_stylesheet) from [`napari.qt`](/api/napari.qt.html) module.

Reviewer: ```suggestion
and [`get_stylesheet`](/api/napari.qt.html#napari.qt.get_stylesheet) functions from the [`napari.qt`](/api/napari.qt.html) module.
```

> Here example for magicgui (but could be easily generalised to qt based widgets):

Reviewer: ```suggestion
Here is a `magicgui` example (but could be easily generalised to native `qt` based widgets):
```

class MyInputDialog(QDialog):
    def __init__(self, parent: QWidget):
        super().__init__(parent)
        self.setWindowTitle("My Input Dialog")
        self.number = QSpinBox()
        self.ok_btn = QPushButton("OK")
        self.cancel_btn = QPushButton("Cancel")
        
        layout = QGridLayout()
        layout.addWidget(QLabel("Number:"), 0, 0)
        layout.addWidget(self.number, 0, 1)
        layout.addWidget(self.ok_btn, 1, 0)
        layout.addWidget(self.cancel_btn, 1, 1)
        self.setLayout(layout)
        
        self.ok_btn.clicked.connect(self.accept)
        self.cancel_btn.clicked.connect(self.reject)
        
class MyWidget(QWidget):
  def __init__(self, viewer: "napari.Viewer"):
        super().__init__()
        self.viewer = viewer
        self.open_dialog = QPushButton("Open dialog")
        self.open_dialog.clicked.connect(self.open_dialog_clicked)
        
    def open_dialog(self):
        # setting parent to self allows the dialog to inherit its 
        # style from the viewer by pass self as argument
        dialog = MyInputDialog(self)  
        dialog.exec_()
        if dialog.result() == QDialog.Accepted:

>             print(dialog.number.value())

Reviewer 2: Should the `open_dialog` function be `open_dialog_clicked`?

Author: yes

```

If there is a particular reason that you need to use a separate window that
inherits from `QWidget`, not `QDialog`, then you could use the `get_current_stylesheet` 
and [`get_stylesheet`](/api/napari.qt.html#napari.qt.get_stylesheet) functions from the [`napari.qt`](/api/napari.qt.html) module.

Here is a `magicgui` example (but could be easily generalised to native `qt` based widgets):

```python
from magicgui import magicgui

from napari.qt import get_stylesheet
from napari.settings import get_settings

def sample_add(a: int, b: int) -> int:
    return a + b

@magicgui
def sample_add(a: int, b: int) -> int:
    return a + b

def change_style():
    sample_add.native.setStyleSheet(get_stylesheet(get_settings().appearance.theme))

get_settings().appearance.events.theme.connect(change_style)

> change_style()

Reviewer 2: Man, it sucks that this is the best way to do this.

@tlambert03 is there a better way to do this?

Reviewer 2: Ah, actually, I see you guys talked about it above.

Should we make an issue about this in the magicgui repo?

Author: I forgot to use the function that I added in this PR which is a wrapper. Could you check the current version?

Author: > Ah, actually, I see you guys talked about it above.

No. Ealier we talked about code that was removed from this PR (embedding mgui widget in dialog).

Reviewer 3: > Man, it sucks that this is the best way to do this.

what specifically sucks

Reviewer 2: What sucks is that you have to write this code to alter the style of your magicgui widget. It would be great if we could encapsulate this behavior with a parameter set by the user somewhere.

I'd probably want to see such a parameter in `Widget.show`. I'm imagining writing something like `sample_add.show(style=get_current_stylesheet())`. Or, if you want to copy the style of some existing widget `widget` it would be slick to write `sample_add.show(style=style_of(widget))`

But I don't know if such a parameter would be best put on that function.

My point is that with all of the abstraction that magicgui makes, it would be sweet if we simplified the passing of a style as well.

Reviewer 3: > What sucks is that you have to write this code to alter the style of your magicgui widget. 

to be clear, you _don't_ have to do that unless you don't pass in a parent widget.  Qt will automagically pass down stylesheets to all child widgets.  So this happens for free for any widgets that are embedded or docked somehow in the napari viewer.  However, once widgets are made that essentially know nothing about the napari parent viewer chain, then this connection is the cost.

So really, this comes down to proper widget parenting.  There _are_ places in magicgui where passing a parent needs to be easier (and, as discussed before, the brand new Dialog is one of those)...

Reviewer 3: i think that confusion is part of the reason that adding this doc is possibly misleading.  I think we probably should generally enforce that, if people want to add widgets to napari, they parent them properly.  (that has implications for more than just styles... it also makes sure that things are cleaned up and garbage collected properly)... whereas this doc kinda makes it look like an equally good idea to make your own independent windows, and style them manually

Author: There are still cases like PartSeg where styling is required (for proper working QtViewer), and there is no option for inheritance. 

So in some way, these docs are useful.

Reviewer 2: > So really, this comes down to proper widget parenting. There _are_ places in magicgui where passing a parent needs to be easier (and, as discussed before, the brand new Dialog is one of those)...

I totally agree that proper parenting is the correct course of action most of the time. And yeah, we wouldn't want people to misuse such a parameter in place of proper parenting.

Is it worth making it easier, though, in cases where we might want a different style than the parent's style? For example, maybe we'd want a different style in napari-imagej dialogs (to mimic the style of Fiji/ImageJ, for example).

But I also don't know that we'd want to promote that...

Author: I agree that better will be to add dialogs to magicgui and promote (`dialog.show(self)`), but I do not have time to make such PR, but I have enough time to write this text to not have a mix of napari window and system styled widgets.
